From 7bfad108942b1c819da07879d98c34af42612d78 Mon Sep 17 00:00:00 2001
From: dianlujitao <dianlujitao@lineageos.org>
Date: Thu, 27 Dec 2018 15:47:07 +0800
Subject: [PATCH 3/4] build: Mount A-only system-as-root image at proper mount
 point

 * By AOSP build system and recovery design, A-only system-as-root
   image should be mounted at /system_root and /system is a symlink
   to /system_root/system.
 * Detect system-as-root status in the proper way.

Change-Id: I2d46feaf660c9bee91a2254e2e45f85fc7c052da
---
 tools/releasetools/edify_generator.py       |  5 ++---
 tools/releasetools/ota_from_target_files.py | 19 ++++++++++---------
 2 files changed, 12 insertions(+), 12 deletions(-)

diff --git a/tools/releasetools/edify_generator.py b/tools/releasetools/edify_generator.py
index 30ff3e5d3..59f9d8831 100644
--- a/tools/releasetools/edify_generator.py
+++ b/tools/releasetools/edify_generator.py
@@ -164,9 +164,8 @@ class EdifyGenerator(object):
            ");")
     self.script.append(self.WordWrap(cmd))
 
-  def RunBackup(self, command, system_path):
-    self.script.append(('run_program("/tmp/install/bin/backuptool.sh", "%s", "%s");' % (
-        command, system_path)))
+  def RunBackup(self, command):
+    self.script.append(('run_program("/tmp/install/bin/backuptool.sh", "%s");' % command))
 
   def ShowProgress(self, frac, dur):
     """Update the progress bar, advancing it over 'frac' over the next
diff --git a/tools/releasetools/ota_from_target_files.py b/tools/releasetools/ota_from_target_files.py
index a5f2d505b..61c73119e 100755
--- a/tools/releasetools/ota_from_target_files.py
+++ b/tools/releasetools/ota_from_target_files.py
@@ -820,7 +820,10 @@ else if get_stage("%(bcb_dev)s") == "3/3" then
   # Dump fingerprints
   script.Print("Target: {}".format(target_info.fingerprint))
 
-  script.AppendExtra("ifelse(is_mounted(\"/system\"), unmount(\"/system\"));")
+  is_system_as_root = target_info.get("system_root_image") == "true"
+  system_mount_point = "/system_root" if is_system_as_root else "/system"
+
+  script.AppendExtra("ifelse(is_mounted(\"{0}\"), unmount(\"{0}\"));".format(system_mount_point))
   device_specific.FullOTA_InstallBegin()
 
   CopyInstallTools(output_zip)
@@ -829,12 +832,11 @@ else if get_stage("%(bcb_dev)s") == "3/3" then
   script.SetPermissionsRecursive("/tmp/install/bin", 0, 0, 0755, 0755, None, None)
 
   if OPTIONS.backuptool:
-    is_system_as_root = script.fstab["/system"].mount_point == "/"
     if is_system_as_root:
-      script.fstab["/system"].mount_point = "/system"
+      script.fstab["/system"].mount_point = system_mount_point
     script.Mount("/system")
-    script.RunBackup("backup", "/system/system" if is_system_as_root else "/system")
-    script.Unmount("/system")
+    script.RunBackup("backup")
+    script.Unmount(system_mount_point)
     if is_system_as_root:
       script.fstab["/system"].mount_point = "/"
 
@@ -881,12 +883,11 @@ else if get_stage("%(bcb_dev)s") == "3/3" then
 
   if OPTIONS.backuptool:
     script.ShowProgress(0.02, 10)
-    is_system_as_root = script.fstab["/system"].mount_point == "/"
     if is_system_as_root:
-      script.fstab["/system"].mount_point = "/system"
+      script.fstab["/system"].mount_point = system_mount_point
     script.Mount("/system")
-    script.RunBackup("restore", "/system/system" if is_system_as_root else "/system")
-    script.Unmount("/system")
+    script.RunBackup("restore")
+    script.Unmount(system_mount_point)
     if is_system_as_root:
       script.fstab["/system"].mount_point = "/"
 
-- 
2.17.1

